File: from-0301-to-0400/0380-insert-delete-getrandom-o(1)/solution.go
type RandomizedSet struct {
    set map[int]bool
    vals []int
}


func Constructor() RandomizedSet {
    set := make(map[int]bool)
    vals := []int{}
    return RandomizedSet{set, vals}
}


func (this *RandomizedSet) Insert(val int) bool {
    if _, ok := this.set[val]; ok {
        return false
    }
    this.set[val] = true
    this.vals = append(this.vals, val)
    return true
}


func (this *RandomizedSet) Remove(val int) bool {
    if _, ok := this.set[val]; !ok {
        return false
    }
    delete(this.set, val)
    valIdx := -1
    for idx, num := range this.vals {
        if num == val {
            valIdx = idx
            break
        }
    }
    this.vals[valIdx] = this.vals[len(this.vals) - 1]
    this.vals = this.vals[:len(this.vals) - 1]
    return true
}


func (this *RandomizedSet) GetRandom() int {
    rndIdx := rand.Intn(len(this.vals))
    return this.vals[rndIdx]
}


/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: from-0301-to-0400/0380-insert-delete-getrandom-o(1)/solution.py
class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.store = set()
        

    def insert(self, val):
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        if val not in self.store:
            self.store.add(val)
            return True
        
        return False
        

    def remove(self, val):
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        :type val: int
        :rtype: bool
        """
        if val in self.store:
            self.store.remove(val)
            return True
        
        return False
        

    def getRandom(self):
        """
        Get a random element from the set.
        :rtype: int
        """
        return random.choice(list(self.store))
        
# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()